s02/2.md

```diff
@@ -10,14 +10,556 @@ Tartalom
+csináljon vele)
+* az adatot típushelyesen tudjuk a hiba fellépési helyétől a handler-hez szállítani
+* ne járjon semmilyen extra (kód/idő/hely) hátránnyal, ha nem használjuk
+* minden kivételt a megfelelő handler kapja el
+* többszálú környezetben is megfelelő
+* ha nincs hiba $\Rightarrow$ ne legyen overheadje (vagy legalábbis minimális legyen)
+    + `if`-ek nem jók, mert a folyamatos kiértékelés órajelet emészt
+    + kivételek megcsinálhatók úgy, hogy majdnem költségmentesek
+    legyenek
+
+#### try-catch
+
+```C++
+try {
+    f();
+    // ...
+}
+catch (T1 e1) { /* handler for T1 */ }
+catch (T2 e2) { /* handler for T2 */ }
+catch (T3 e3) { /* handler for T3 */ }
+```
+
+* `throw` - bármi dobható, de az értelmes: `std:exception` és annak leszármazottjai
+* `catch(T e)` - `T` típusú érték handlere
+    1. Elkapja a kivételt, ha az `T` típusú
+    2. vagy annak leszármazottja
+    3. illetve pointer vagy referencia és a hivatkozott értékre fennáll 1) vagy 2)
+* Nem szabad `new`-val exception-t létrehozni $\Rightarrow$ memóriaszivárgás
+
+#### Hierarchia
+
+* Az öröklődést használjuk kivételek csoportosítására
+* Az általánosabb handler elkapja a speciálisabb exceptiont
+* A `catch` ágak a megadott sorrendben értékelődnek ki
+* ezért figyelni kell a handlerek általánosságát és sorrendjét
+    + A speciálisabb kerüljön felülre és az általánosabb alulra
+
+Továbbra sem ajánlott `new`-val kivételt létrehozni. Ha a dianamikus típussal akarunk játszani,
+inkább használjuk a következő megoldást:
+
+```C++
+struct ExceptionBase{
+    virtual void raise() { throw *this; }
+    virtual ~ExceptionBase() {}
+};
+
+struct ExceptionDerived : ExceptionBase{
+    virtual void raise() { throw *this; }
+};
+
+void foo(ExceptionBase& e){
+    e.raise(); // Uses dynamic type of e while raising an exception.
+}
+int main (void){
+    ExceptionDerived e;
+    try {
+        foo(e);
+    }catch (ExceptionDerived& e) {
+        ...
+    }catch (...) {
+        ...
+    }
+}
+```
+
+#### Kivételkezelés és osztályok
+
+Kérdéses esetek: konstruktor, destruktor
+
+* Konstruktor
+    + Ha a konstruktor dob, terület le lett foglalva, de a pointer nem lett beállítva $\Rightarrow$ nincs gond,
+    a rendszer deallokálja
+    + De: a konstruktoron belül lefogalt területet a destruktor tudja felszabadítani. A destruktort viszont nem lehet 
+    meghívni mert az objektum létre se jött rendesen $\Rightarrow$ probléma
+    ```C++
+    class X
+    {
+        public:
+            X(int i) { p = new char[i]; init(); }
+            ~X() { delete [] p; }       // must not throw exception
+        private:
+            void init() { ... throw ... }   // BAD: destructor won't run !
+            char *p;                        // constructor was not completed
+    };
+    ```
+    
+    + Ha tagváltozó inicializálása dob hibát, akkor dob a konstruktor is
+    ```C++
+    class X
```

Comment:
(\newpage az fontos)

\newpage  

    + Ha tagváltozó inicializálása dob hibát, akkor dob a konstruktor is

```cpp
class X
{
    public:
        X()
        {
            throw std::runtime_error("Exception thrown from member.");
        }
};

class Y
{
    public:
        Y()
        try
            : x()   // You can have try-catch in initializer lists too
        { }
        catch( std::runtime_error& e )
        {
            throw e;  
        }

    private:
        X x;
};

int main()
{
    try
    {
        Y y;
        return 0;
    }
    catch ( std::runtime_error& e)
    {
        std::cerr << "exception: " << e.what() << std::endl;
    }
}
```